ui/src/app/pages/analysis/runtime-panel.tsx

```diff
@@ -79,178 +86,226 @@ interface State {
-          promise: promise,
-          operation: 'get',
-          aborter: this.aborter
-        });
-        runtime = await promise;
-      } catch (e) {
-        // 404 is expected if the runtime doesn't exist, represent this as a null
-        // runtime rather than an error mode.
-        if (e.status !== 404) {
-          error = true;
-        }
-      }
-      markRuntimeOperationCompleteForWorkspace(this.props.workspace.namespace);
-      this.setState({
-        runtime,
-        error,
-        loading: false
-      });
-    }
+const DiskSizeSelection = ({onChange, selectedDiskSize, currentRuntime}) => {
+  const {dataprocConfig, gceConfig} = currentRuntime;
+  const masterDiskSize = !!dataprocConfig ? dataprocConfig.masterDiskSize : gceConfig.bootDiskSize
 
-    render() {
-      const {runtimeOps, workspace} = this.props;
-      const {loading, error, runtime} = this.state;
+  return <div>
+    <label htmlFor='runtime-disk'
+          style={{marginRight: '.25rem'}}>Disk (GB)</label>
+      <InputNumber id='runtime-disk'
+                //  disabled={true}
+                showButtons
+                decrementButtonClassName='p-button-secondary'
+                incrementButtonClassName='p-button-secondary'
+                value={selectedDiskSize || masterDiskSize}
+                inputStyle={{padding: '.75rem .5rem', width: '2rem'}}
+                onChange={({value}) => onChange(value === masterDiskSize ? null : value)}
+                min={50 /* Runtime API has a minimum 50GB requirement. */}/>
+  </div>
+}
 
-      const activeRuntimeOp: RuntimeOperation = runtimeOps.opsByWorkspaceNamespace[workspace.namespace];
+const updateRuntime = async ({workspaceNamespace, currentRuntime, selectedDiskSize, selectedMachine}) => {
+  const {dataprocConfig: {masterDiskSize, masterMachineType}} = currentRuntime;
+  const nextMachineType = selectedMachine && selectedMachine.name;
+  const pollAbortSignal = new AbortController();
 
-      if (loading) {
-        return <Spinner style={{width: '100%', marginTop: '5rem'}}/>;
-      } else if (error) {
-        return <div>Error loading compute configuration</div>;
-      } else if (!runtime) {
-        // TODO(RW-5591): Create runtime page goes here.
-        return <React.Fragment>
-          <div>No runtime exists yet</div>
-          {activeRuntimeOp && <hr/>}
-          {activeRuntimeOp && <div>
-            <ActiveRuntimeOp operation={activeRuntimeOp.operation} workspaceNamespace={workspace.namespace}/>
-          </div>}
-        </React.Fragment>;
-      }
+  await runtimeApi().deleteRuntime(workspaceNamespace);
+
+  await LeoRuntimeInitializer.initialize({
+    workspaceNamespace,
+    dataprocConfig: {
+      masterMachineType: nextMachineType || masterMachineType,
+      masterDiskSize: selectedDiskSize || masterDiskSize
+    },
+    pollAbortSignal: pollAbortSignal.signal,
+    onStatusUpdate: status => status === 'Running' && pollAbortSignal.abort()
```

Comment:
Ah - thanks - I will remove this.

I had the aborter setup on unmount, but I don't think this is what we want - the UX was unpredictable. If a user chooses to change their runtime and navigates away, refreshes, etc... Then they lose the customization of their runtime. The next time they go in to the notebook they will get the default runtime, causing them to have to customize all over again and essentially wait for 2 creates and 1 delete, which can be lengthy.

I don't think it would be unreasonable or out of the ordinary for a user to refresh the page during a long running operation.

Reply:
Yes - I agree we need this scenario to work properly, I'm not suggesting that you break it. But this specific implementation doesn't work as-is, since it will introduce nasty bugs when you switch between workspaces in the application.

The motivation for the `runtimeOperationStore` was to address this exact issue. My proposal was to use that to track this long-running operation so that you can restore when rerendering the panel. If you need to attach the pending runtime config value so you can restore the state of the panel - that's definitely within reason.

Reply:
The problem here is that we're using the LeoInitializer, which is having side-effects I think there are two potential ways I can think to move forward:

1. Instead of `currentRuntimeStore`, switch to a store per runtime (i.e. per workspace), similar to the runtimeOpStore
2. Somehow invalidate the LeoRuntimeInitializer updates to the currentRuntimeStore when the store goes out of context. Possibly disallow it to make an update if the runtime name/namespace don't match the previous one - only allow switches via workspace-page.ts

Reply:
In RW-5409 what I have right now is a `currentRuntimeStore` that clears the store and creates a new `LeoRuntimeInitializer` whenever the workspace context switches. I'm going to have a PR up for that in a little while. If a per-workspace store would make your life easier, I can make that store in to a `workspaceRuntimeStore` pretty easily.